wfe2/wfe.go

```diff
@@ -1198,15 +1202,62 @@ func (wfe *WebFrontEndImpl) Certificate(ctx context.Context, logEvent *web.Reque
 		return
 	}
 
-	response.Header().Set("Content-Type", "application/pem-certificate-chain")
-	response.WriteHeader(http.StatusOK)
-
-	pem := pem.EncodeToMemory(&pem.Block{
+	leafPEM := pem.EncodeToMemory(&pem.Block{
 		Type:  "CERTIFICATE",
 		Bytes: cert.DER,
 	})
-	// TODO(#3291): Serve entire chain instead of just end-entity certificate.
-	if _, err = response.Write(pem); err != nil {
+
+	var responsePEM []byte
+
+	// If the WFE is configured with CertificateChains, construct a chain for this
+	// certificate using its AIA Issuer URL.
+	if len(wfe.CertificateChains) > 0 {
+		parsedCert, err := x509.ParseCertificate(cert.DER)
+		if err != nil {
+			// If we can't parse one of our own certs there's a serious problem
+			wfe.sendError(response, logEvent, probs.ServerInternal(
+				fmt.Sprintf(
+					"unable to parse Boulder issued certificate with serial %#v",
+					serial),
+			), err)
+			return
+		}
+
+		// NOTE(@cpu): Boulder assumes there will only be **ONE** AIA issuer URL
+		// configured in the CA signing profile. At present this is not enforced by
+		// the CA, but should be. See
+		//  https://github.com/letsencrypt/boulder/issues/3374
+		aiaIssuerURL := parsedCert.IssuingCertificateURL[0]
+		if chain, ok := wfe.CertificateChains[aiaIssuerURL]; ok {
+			responsePEM = append([]byte(chain), leafPEM...)
```

Comment:
See comment above: The end-entity certificate needs to come first. Also, it seems like `bytes.Join(leafPem, chain)` would be good here, and slightly clearer. It would require a leading newline instead of a trailing newline in `chain`, as you mentioned in your previous comment.

Reply:
I disagree that `bytes.Join()` is clearer. It makes it seem like `chain` is a delimiter. We just want to concatenate the leaf bytes with the chain bytes, `append` expresses that clearest IMO.

I ended up changing the `certificateChains` type from `map[string]string` to `map[string][]byte` since there isn't any reason to map the PEM to/from strings. With that change & the above rationale for using `append` do you still want to see `bytes.Join` used?

Reply:
Oops, you're right the `bytes.Join(leafPem, chain)` makes it seem like chain is a delimeter (and also wouldn't compile)! I guess I was thinking of `bytes.Join([][]byte{leafPem, chain}, "\n")`, but now that I type it out it doesn't seem all that much better. :-)